File: lab/c/textos/contagem_de_frase.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int contar_palavras1(const char *texto) {
    int cont = 0;
    int dentro_palavra = 0;

    for (int i = 0; texto[i] != '\0'; i++) {
        if (!isspace((unsigned char)texto[i]) && !dentro_palavra) {
            dentro_palavra = 1;
            cont++;
        } else if (isspace((unsigned char)texto[i])) {
            dentro_palavra = 0;
        }
    }
    return cont;
}

int contar_palavras2(const char *texto) {
    int cont = 0;
    int i = 0;

    // pula espaços iniciais
    while (isspace((unsigned char)texto[i])) i++;
    if (texto[i] == '\0') return 0;

    cont = 1;
    for (; texto[i] != '\0'; i++) {
        if (isspace((unsigned char)texto[i])) {
            // verifica se o próximo caractere não é espaço
            while (isspace((unsigned char)texto[i])) i++;
            if (texto[i] != '\0')
                cont++;
        }
    }
    return cont;
}

int main(void) {
    const char *testes[] = {
        "Eu gosto de estudar programação",
        "Olá",
        "Olá Mundo",
        "",
        "   "
    };

    int total = sizeof(testes) / sizeof(testes[0]);

    for (int i = 0; i < total; i++) {
        printf("%d\n", contar_palavras1(testes[i]));
        printf("%d\n", contar_palavras2(testes[i]));
        printf("%d\n\n", contar_palavras1(testes[i])); // sem regex em C
    }

    return 0;
}

File: lab/javascript/textos/contagem_de_frase.js
function contarPalavras1(texto) {
  return texto.trim().split(/\s+/).filter(Boolean).length;
}

function contarPalavras2(texto) {
  texto = texto.trim();
  if (texto.length === 0) return 0;

  let cont = 1;
  let anterior = "";

  for (const ch of texto) {
    // considera o espaço como separador de palavras
    if (ch === " " && anterior !== " ") {
      cont++;
    }
    anterior = ch;
  }
  return cont;
}

function contarPalavras3(texto) {
  // \w+ encontra blocos de caracteres alfanuméricos (Unicode)
  const palavras = texto.trim().match(/\w+/gu);
  return palavras ? palavras.length : 0;
}

// Testes
const testes = [
  "Eu gosto de estudar programação",
  "Olá",
  "Olá Mundo",
  "",
  "   "
];

for (const t of testes) {
  console.log(contarPalavras1(t));
  console.log(contarPalavras2(t));
  console.log(contarPalavras3(t));
  console.log();
}
